### PR TITLE
[DRAFT] Feature/task18

### PR DESCRIPTION
**Pull request by PavelNikolaichev**
[WIP] Authenticate user after they were registered successfully. Allow only 3 login attempts. If a user failed, block him by IP for 15 minutes. Log this fact (log attacker IP-address and email, start and end blocking period) Only authenticated users should have access to File upload form and be able to upload files. Log successful upload operation (message format: {date}: {actual filename} - {filesize} - {archive name} - {archive filesize}) Log all errors and exceptions in separate storage. Implement 'Remember me' functionality and save authenticated user for 1 week.

### CODE REVIEW

File: src/app/core/SessionAuthenticatedUser.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    private array $user;
+
+    public function __construct(private LoginRepository $repository)
+    {
+        if (isset($_SESSION['user'], $_SESSION['email'])) {
+            $this->user = ['email' => $_SESSION['email'], 'name' => $_SESSION['user']];
+        }
+
+        if (isset($_COOKIE['token'])) {
+            $user = $this->repository->fetchByToken($_COOKIE['token']);
```
Comment (PavelNikolaichev) [inline src/app/core/SessionAuthenticatedUser.php:16]: Add check if obj have been fetched. Also check if session have been started.

File: src/app/core/SessionAuthenticatedUser.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            $_SESSION['user'] = $user->first_name;
+            $_SESSION['email'] = $user->email;
+
+            $this->user = ['email' => $user->email, 'name' => $user->first_name];
+        }
+    }
+
+    public function getUser(): ?array
+    {
+        return $this->user ?? null;
```
Comment (PavelNikolaichev) [inline src/app/core/SessionAuthenticatedUser.php:39]: Implement a DTO/Interface for user obj.

File: src/app/migrations/m0005_migrate.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+class m0005_migrate
+{
+    public function up(PDO $pdo): void
+    {
+        $pdo->exec('create table attempts (
+            id     int auto_increment unique,
+            ip  int(10) unsigned  not null unique,
+            email   varchar(255) not null,
+            password varchar(255) not null,
```
Comment (PavelNikolaichev) [inline src/app/migrations/m0005_migrate.php]: Remove password field

Comment (PavelNikolaichev) [reply src/app/migrations/m0005_migrate.php]: Remove email too.

File: src/app/core/Database/AttemptsRepository.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    public function incrementAttempt(): array
+    {
+        $data = $this->queryBuilder
+            ->fetch('attempts')
+            ->select(['*'])
+            ->where('ip', '=', ip2long($this->getIP()))
+            ->get();
+
```
Comment (pavel-krulov) [inline src/app/core/Database/AttemptsRepository.php]: Insert into aaa(ip, attempts)values ('1.2.3.4',1) on duplicate key update (attempts = attempts + 1);

**Pull request merged**